Add Timeout at Replica Sync Remote Comms

vazois opened the pull request:
This PR adds missing timeout for over the network communications between primary and secondary.

TalZaccai commented on libs/cluster/Server/Replication/PrimaryOps/ReplicaSyncSession.cs:
```diff
@@ -278,7 +278,7 @@ public async Task<bool> SendCheckpoint()
                     localEntry.ToByteArray(),
                     beginAddress,
-                    checkpointAofBeginAddress).ConfigureAwait(false);
+                    checkpointAofBeginAddress).WaitAsync(timeout, cts.Token);
```
Would it make sense to add the `.ConfigureAwait(false)` after `WaitAsync`?

TalZaccai commented on libs/cluster/Server/Replication/PrimaryOps/ReplicaSyncSession.cs:
```diff
@@ -571,7 +572,7 @@ private unsafe (SectorAlignedMemory, int) ReadInto(IDevice device, ulong address
             else
                 device.ReadAsync(segmentId, address, (IntPtr)pbuffer.aligned_pointer, (uint)numBytesToRead, IOCallback, null);
-            semaphore.Wait();
+            semaphore.Wait(clusterProvider.clusterManager.clusterTimeout, cts.Token);
```
Should this be async as well?

vazois replied:
I had to remove the unsafe specifier for this method, but I think it should be ok to make it WaitAsync.

TalZaccai commented on libs/cluster/Server/Replication/PrimaryOps/ReplicaSyncSession.cs:
```diff
@@ -559,19 +564,22 @@ private async Task SendObjectFiles(GarnetClientSession gcs, Guid token, Checkpoi
+                else
+                    device.ReadAsync(segmentId, address, (IntPtr)pbuffer.aligned_pointer, (uint)numBytesToRead, IOCallback, null);
+            }
+            await semaphore.WaitAsync(clusterProvider.clusterManager.clusterTimeout, cts.Token);
```
Can add ConfigureAwait(false) here too?

priyanjgupta commented:
Sorry for this basic question I'm still getting familiar with the replication internals.
**Scenario**: What happens if a timeout occurs while the secondary is actively processing a file segment?
What happens:
**Primary side**: Throws TimeoutException → Exception is caught in the general handler → Returns false (marking the operation as failed)
**Secondary side**: Continues processing the segment → May successfully complete the write operation
Can this create any inconsistencies ? 
Is there a protocol-level mechanism that handles such mid-operation timeouts to prevent inconsistencies?
And if not, should the replication logic be concerned about such scenarios?
From my understanding, this might not lead to issues because a timeout causes `NetworkClusterInitiateReplicaSync` to fail, and the secondary would reset it's state. If replication is retried, it would reprocess the file from the beginning, which could lead to duplicate writes.
**_Could you please confirm whether the secondary safely rewrites the data in such cases, rather than rejecting it or ending up in an inconsistent state?Just want to validate that my understanding of the safety mechanism is correct._**
Thanks

vazois commented:
> Sorry for this basic question I'm still getting familiar with the replication internals. **Scenario**: What happens if a timeout occurs while the secondary is actively processing a file segment? What happens: **Primary side**: Throws TimeoutException → Exception is caught in the general handler → Returns false (marking the operation as failed) **Secondary side**: Continues processing the segment → May successfully complete the write operation Can this create any inconsistencies ? Is there a protocol-level mechanism that handles such mid-operation timeouts to prevent inconsistencies? And if not, should the replication logic be concerned about such scenarios? From my understanding, this might not lead to issues because a timeout causes `NetworkClusterInitiateReplicaSync` to fail, and the secondary would reset it's state. If replication is retried, it would reprocess the file from the beginning. **_Could you please confirm whether the secondary safely rewrites the data in such cases, rather than rejecting it or ending up in an inconsistent state?Just want to validate that my understanding of the safety mechanism is correct._** Thanks

This is a good observation and yes, the primary should rewrite the contents of the checkpoint because it starts rewriting from scratch. I added a test that demonstrates this.
When primary fails for any reason, the secondary should receive a response and cleanup any acquired locks and resources etc.
In that case, when re-initiating the sync path the files associated with the checkpoint will be overwritten.
If the secondary is unreachable, I added a timeout to the initiateReplicaSyncTask in order to avoid the secondary being stuck.

TalZaccai commented on libs/cluster/Server/Replication/ReplicaOps/ReceiveCheckpointHandler.cs:
```diff
@@ -65,6 +74,10 @@ public void ProcessFileSegments(int segmentId, Guid token, CheckpointFileType ty
             WriteInto(writeIntoCkptDevice, (ulong)startAddress, data, data.Length, segmentId);
+
+#if DEBUG
+            ExceptionInjectionHelper.WaitOnClear(ExceptionInjectionType.Replication_Timeout_On_Receive_Checkpoint).GetAwaiter().GetResult();
```
`.ConfigureAwait(false)` before `GetAwaiter()`?

TalZaccai commented on libs/cluster/Server/Replication/ReplicaOps/ReceiveCheckpointHandler.cs:
```diff
@@ -77,27 +90,37 @@ public void ProcessFileSegments(int segmentId, Guid token, CheckpointFileType ty
+                    device.WriteAsync((IntPtr)pbuffer.aligned_pointer, segmentId, address, (uint)numBytesToWrite, IOCallback, null);
 
-            pbuffer.Return();
+                _ = writeCheckpointSemaphore.Wait(replicaSyncTimeout, cts.Token);
```
Why not do a WaitAsync here?

TalZaccai commented on libs/cluster/Server/Replication/ReplicaOps/ReceiveCheckpointHandler.cs:
```diff
@@ -77,27 +90,37 @@ public void ProcessFileSegments(int segmentId, Guid token, CheckpointFileType ty
+            catch (Exception ex)
+            {
+                logger?.LogError(ex, $"{nameof(ReceiveCheckpointHandler)}.WriteInto");
+                throw new Exception(ex.Message);
```
Why not just rethrow ex?

vazois replied:
You are right, the exception should be captured at ProcessMessage level. I will make the change.

TalZaccai commented on libs/common/ExceptionInjectionHelper.cs:
```diff
@@ -87,5 +87,20 @@ public static async Task WaitOnCondition(ExceptionInjectionType exceptionType)
+        /// </summary>
+        /// <param name="exceptionType"></param>
+        /// <returns></returns>
+        public static async Task WaitOnClear(ExceptionInjectionType exceptionType)
```
nit: Not sure if we have a proper naming convention for async methods, but should probably end with `Async`

TalZaccai commented on libs/common/ExceptionInjectionHelper.cs:
```diff
@@ -87,5 +87,20 @@ public static async Task WaitOnCondition(ExceptionInjectionType exceptionType)
+        /// <returns></returns>
+        public static async Task WaitOnClear(ExceptionInjectionType exceptionType)
+        {
+            var flag = ExceptionInjectionTypes[(int)exceptionType];
```
nit: Maybe a bit cleaner?
```cs
var flag = ExceptionInjectionTypes[(int)exceptionType];
while (flag)
{
    await Task.Yield();
    flag = ExceptionInjectionTypes[(int)exceptionType];
}
```

vazois replied:
actually, I need to remove if because it is redundant

TalZaccai reviewed (changes requested):
Added a few comments
